### PR TITLE
Use CheckpointManager for checkpointing

### PR DESCRIPTION
**Pull request by jonb377**
[CheckpointManager](https://github.com/pytorch/xla/blob/46643801a5da36106c346d0434387c29d3a7818d/torch_xla/experimental/distributed_checkpoint/manager.py#L39) provides an interface for distributed checkpointing with SPMD.

This change integrates CheckpointManager into the standard HF checkpoint control path. This is an early integration, and the final form of the integration is subject to change significantly. It will provide a convenient way for end-to-end testing of checkpointing and convergence.

Tested by checkpointing locally with ~`--use_checkpoint_manager --save_strategy epoch`~ `--checkpoint_manager_path gs://pytorch-spmd-benchmark-profile/devbox/1700201148` with a 2B Llama config for 1 epoch, achieving a loss of `10.2381`. Then, re-ran for 2 epochs, and observed the checkpoint resumed from the end of the first epoch, resulting in a loss of `5.0476` (normalizes to approximately `10.0952`, which shows the loss resumed decreasing):

```
...
[INFO|trainer.py:1754] 2023-11-09 04:58:09,244 >>   Continuing training from checkpoint, will skip to saved global_step
[INFO|trainer.py:1755] 2023-11-09 04:58:09,244 >>   Continuing training from epoch 1
[INFO|trainer.py:1756] 2023-11-09 04:58:09,244 >>   Continuing training from global step 21
[INFO|trainer.py:1758] 2023-11-09 04:58:09,244 >>   Will skip the first 1 epochs then the first 0 batches in the first epoch.
...
```

This change depends on https://github.com/pytorch/xla/pull/5770 for a fix in async checkpointing.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -519,8 +522,22 @@ def __init__(
...
+            self.chkpt_manager = CheckpointManager(
+                path=self.args.output_dir,
+                save_interval=float('inf'),  # Defer to the HF save strategy
+                max_to_keep=self.args.save_total_limit or 0
+            )
+
         default_callbacks = DEFAULT_CALLBACKS + get_reporting_integration_callbacks(self.args.report_to)
         callbacks = default_callbacks if callbacks is None else default_callbacks + callbacks
+        if self.args.use_checkpoint_manager:
+            callbacks.append(CheckpointManagerCallback(self.chkpt_manager))
```
Comment (alanwaketan) [inline src/transformers/trainer.py:540]: Can you tell me more about what these callbacks do?

Comment (jonb377) [reply src/transformers/trainer.py:540]: These callbacks are invoked at various times in training, e.g. `on_step_end`, `on_train_start`, `on_epoch_end`. They're meant to modify the control state or take action at certain times, which for us would mean modifying `control.should_save` based on `chpkt_manager.should_save()` each step (`on_step_end`).

Since the HF control manages checkpoint frequency, the only time `chpkt_manager.should_save` would override the value is when a preemption is caught.

Comment (alanwaketan) [reply src/transformers/trainer.py:540]: Thanks for the explanation.

**Review by alanwaketan — APPROVED** [review]

**Comment by jonb377** [comment]
cc @alanwaketan, I changed the flag to be a path rather than bool. In the trainer, the output-dir pretty much has to be a local path, which made a lot of special casing for CheckpoingManager. Now, output-dir can remain a local path and checkpoint_manager_path can target GCS.

**Comment by alanwaketan** [comment]
> cc @alanwaketan, I changed the flag to be a path rather than bool. In the trainer, the output-dir pretty much has to be a local path, which made a lot of special casing for CheckpoingManager. Now, output-dir can remain a local path and checkpoint_manager_path can target GCS.

Awesome changes!
